### PR TITLE
New schemas for Retail

### PR DESCRIPTION
**Pull request by ybyzek**
## Problem

New datasets:

- transactions
- credit cards
- stores

Modified datasets:

- purchases: schema existed but wasn't defined as a quickstart, so it could have never been used before.  Therefore breaking changes should be ok

##### Does this solution apply anywhere else?
- [ ] yes
- [ ] no

##### Testing done:
- [ ] Unit tests
- [ ] Integration tests
- [ ] System tests
- [ ] Manual tests

- [x] Run `./connector-test.sh` to validate (requires `mvn` and local CP install)

Sample output:

```
credit_cards:
The local commands are intended for a single-node development environment only,
NOT for production usage. https://docs.confluent.io/current/cli/index.html

7	{"card_id":7,"card_number":5555555555554444,"cvv":273,"expiration_date":"10/27"}
8	{"card_id":8,"card_number":5105105105105100,"cvv":663,"expiration_date":"10/27"}
18	{"card_id":18,"card_number":4000000760000002,"cvv":922,"expiration_date":"04/27"}
8	{"card_id":8,"card_number":5105105105105100,"cvv":663,"expiration_date":"10/27"}
9	{"card_id":9,"card_number":4111111111111111,"cvv":822,"expiration_date":"10/26"}
Processed a total of 5 messages

stores:
The local commands are intended for a single-node development environment only,
NOT for production usage. https://docs.confluent.io/current/cli/index.html

3	{"store_id":3,"city":"Sacramento","state":"CA"}
4	{"store_id":4,"city":"Austin","state":"TX"}
5	{"store_id":5,"city":"Boston","state":"MA"}
1	{"store_id":1,"city":"Raleigh","state":"NC"}
3	{"store_id":3,"city":"Sacramento","state":"CA"}
Processed a total of 5 messages

transactions:
The local commands are intended for a single-node development environment only,
NOT for production usage. https://docs.confluent.io/current/cli/index.html

1	{"transaction_id":1,"card_id":7,"user_id":"User_1","purchase_id":0,"store_id":2}
2	{"transaction_id":2,"card_id":16,"user_id":"User_6","purchase_id":1,"store_id":3}
3	{"transaction_id":3,"card_id":10,"user_id":"User_5","purchase_id":2,"store_id":1}
4	{"transaction_id":4,"card_id":9,"user_id":"User_5","purchase_id":3,"store_id":3}
5	{"transaction_id":5,"card_id":2,"user_id":"User_3","purchase_id":4,"store_id":6}
Processed a total of 5 messages

purchases:
The local commands are intended for a single-node development environment only,
NOT for production usage. https://docs.confluent.io/current/cli/index.html

0	{"id":0,"item_type":"shirt","quantity":6,"price_per_unit":48.14}
1	{"id":1,"item_type":"clippers","quantity":5,"price_per_unit":11.81}
2	{"id":2,"item_type":"scarf","quantity":5,"price_per_unit":39.94}
3	{"id":3,"item_type":"clippers","quantity":8,"price_per_unit":21.87}
4	{"id":4,"item_type":"boots","quantity":9,"price_per_unit":41.41}
Processed a total of 5 messages


## Release Plan
<!--- Describe the release plan for this feature. -->
<!-- Are you backporting or merging to master? -->
<!-- If you are reverting or rolling back, is it safe? -->

### CODE REVIEW

File: src/main/resources/product.avro [hunk 1]
```diff
@@ -28,12 +43,13 @@
                     }
                  }},
                  {"name": "price", "type": {
-                    "type": "double",
+                    "type": "long",
```
[the remark this answers is not in the document]
Comment (ybyzek) [reply src/main/resources/product.avro:46]: @C0urante I'll move this to v2

File: src/main/resources/credit_cards.avro [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+                { "card_id": 24,
+                  "card_number": 40000038000004468,
+                  "cvv": 723,
+                  "expiration_date": "01/27"
+                },
+                { "card_id": 25,
+                  "card_number": 40000535600000118,
+                  "cvv": 723,
+                  "expiration_date": "05/27"
+                }
```
Comment (gharris1727) [inline src/main/resources/credit_cards.avro:149]: All but these last two pass the https://en.wikipedia.org/wiki/Luhn_algorithm checksum, but these are too long to be CCs. WDYT about having at least one that's the right length but fails the checksum?
```suggestion
                { "card_id": 24,
                  "card_number": 4000003800000448,
                  "cvv": 723,
                  "expiration_date": "01/27"
                },
                { "card_id": 25,
                  "card_number": 40000535600000118,
                  "cvv": 723,
                  "expiration_date": "05/27"
                }
```

Comment (ybyzek) [reply src/main/resources/credit_cards.avro]: @gharris1727 great suggestion.  Please see latest commit:

- Removed last 2 CCs
- Added 1 CC that fails the checksum (last one)

File: src/main/resources/credit_cards.avro [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+            ]
```
[the remark this answers is not in the document]
Comment (C0urante) [reply src/main/resources/credit_cards.avro]: IIRC the options file feature works exactly the same as the inline options feature, it just allows you to extract the list of options into a separate, reusable location. I think that that should satisfy the constraints you outline here as long as they're already satisfied by the inline options list.

Not a huge deal either way, just thought it might be useful if you want the same list of credit cards to be reusable across different generated data sets.

[the remark this answers is not in the document]
Comment (ybyzek) [reply src/main/resources/credit_cards.avro]: @C0urante I don't think there's a use case that would require that change.  I'm fine with it being embedded as it is now, if you are.

[the remark this answers is not in the document]
Comment (C0urante) [reply src/main/resources/credit_cards.avro]: Sounds good 👍

File: src/main/resources/transactions.avro [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+                        "range": {
+                            "min": 1,
+                            "max": 24
+                        }
+                    }
+                }},
+                {"name": "user_id", "type": {
+                    "type": "string",
+                    "arg.properties": {
+                        "regex": "User_[1-9]{0,1}"
```
Comment (bellemare) [inline src/main/resources/transactions.avro:26]: @bbejeck - do we want a broader user range for high partition counts keyed on userId? can we have a broader user range?

**Review by bellemare — COMMENTED** [review]
Couple points to discuss - see review.

File: src/main/resources/productv2.avro [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+                }},
+                {"name": "description", "type": {
+                    "type": "string",
+                    "arg.properties": {
+                      "iteration": {
+                          "start": 0
+                      }
+                    }
+                 }},
+                 {"name": "price", "type": {
```
Comment (ybyzek) [inline src/main/resources/productv2.avro:45]: @C0urante just wanted to check in with you on this part.  `price` would ideally be of the dollar form as a `decimal` with 2 digits after the decimal point (unlike the `long` that it is now).  To that end, I tried a few different variations as demonstrated in https://github.com/confluentinc/avro-random-generator/blob/master/src/test/resources/test-schemas/decimals.json but they resulted in varying errors.  If you happen to have the cycles, would you be able to provide a working example of creating a realistic `price` field as a decimal?  

If we can't get that working, there is another option which is to construct it as a `string` like:

```
"regex":"[1-9][0-9]{0,1}[.][0-9]{2,2}"
```

Or we just leave it as it is...

Comment (C0urante) [reply src/main/resources/productv2.avro:45]: Hmm... would that be a decimal with a scale of 2? IIRC the way Avro represents decimal types is with an unscaled value multiplied by `10^(-1 * scale)`, so a scale of 2 would guarantee at most two digits after the decimal point.

Comment (ybyzek) [reply src/main/resources/productv2.avro:45]: > a scale of 2 would guarantee at most two digits after the decimal point.

Yes, this is the desired outcome.  But all syntaxes that I tried resulted in errors.  For example:

```
[11:44:32] ~/git/kafka-connect-datagen/src/main/resources(retail) ✗: git diff
diff --git a/src/main/resources/productv2.avro b/src/main/resources/productv2.avro
index 002c579..98522ab 100644
--- a/src/main/resources/productv2.avro
+++ b/src/main/resources/productv2.avro
@@ -42,14 +42,14 @@
                       }
                     }
                  }},
-                 {"name": "price", "type": {
-                    "type": "long",
-                     "arg.properties": {
-                        "range": {
-                            "min": 2,
-                            "max": 20
-                        }
-                     }
-                  }}
+   {
+     "name": "price",
+     "type": {
+       "type": "bytes",
+       "logicalType": "decimal",
+       "precision": 4,
+       "scale": 2
+     }
+   }
         ]
 }
```

Results in:

```
[2021-09-13 11:44:36,828] ERROR Unable to parse the provided schema (io.confluent.kafka.connect.datagen.ConfigUtils:60)
org.apache.avro.SchemaParseException: com.fasterxml.jackson.core.io.JsonEOFException: Unexpected end-of-input: expected close marker for Object (start marker at [Source: (BufferedInputStream); line: 1, column: 1])
```

Comment (C0urante) [reply src/main/resources/productv2.avro:45]: Hmm... I was able to run `arg` (the Avro Random Generator CLI) locally against the latest `productv2` schema with this change:

```
-                    "type": "long",
-                     "arg.properties": {
-                        "range": {
-                            "min": 2,
-                            "max": 20
-                        }
-                     }
+                    "type": "bytes",
+                    "logicalType": "decimal",
+                    "scale": 2,
+                    "precision": 4
```

I can provide the complete schema if that'd be helpful, LMK.

Comment (C0urante) [reply src/main/resources/productv2.avro:45]: Ah, weird. Even though the `arg` CLI works, the connector tests fail with the `SchemaParseException` you provided.

Comment (C0urante) [reply src/main/resources/productv2.avro:45]: Okay, even weirder--seeing those same errors in the logs for builds on the latest for this branch, even without the changes to the `productV2` schema. They're not causing the build to fail, though.

Comment (C0urante) [reply src/main/resources/productv2.avro:45]: Looks like the "real" cause of the failure once the `productV2` schema is changed to use the logical decimal type is this:
```
java.lang.AssertionError
	at io.confluent.kafka.connect.datagen.DatagenTaskTest.assertRecordsMatchSchemas(DatagenTaskTest.java:270)
	at io.confluent.kafka.connect.datagen.DatagenTaskTest.generateAndValidateRecordsFor(DatagenTaskTest.java:235)
	at io.confluent.kafka.connect.datagen.DatagenTaskTest.shouldGenerateFilesForProductv2Quickstart(DatagenTaskTest.java:137)
```

Seems like the logic in `DatagenTaskTest::isConnectInstance` might need to be updated to handle `ByteBuffer` instances, based on [this arg logic](https://github.com/confluentinc/avro-random-generator/blob/f83f891c126aed54de892670cd7efbf6fb2c13a9/src/main/java/io/confluent/avro/random/generator/Generator.java#L1100-L1110).

Comment (ybyzek) [reply src/main/resources/productv2.avro:45]: @C0urante can you please provide more guidance on the fix?  Why isn't [this case](https://github.com/confluentinc/kafka-connect-datagen/blob/master/src/test/java/io/confluent/kafka/connect/datagen/DatagenTaskTest.java#L257-L258) handling the logic?

Comment (C0urante) [reply src/main/resources/productv2.avro:45]: So I initially thought that that case wasn't sufficient because the Avro Random Generator produces `ByteBuffer` instances instead of `byte[]` instances when producing logical decimal data. However, it's actually slightly more complicated than that, since the connector itself then converts the Avro data produced by the generator to the Kafka Connect format using the `AvroData` class from the Avro converter. That conversion causes that `ByteBuffer` to be turned into a `BigDecimal` instance (see the [AvroData](https://github.com/confluentinc/schema-registry/blob/ff34ad373a3a59c9da812b93e04cecbd2646f1c7/avro-data/src/main/java/io/confluent/connect/avro/AvroData.java#L212-L214) and [Connect Decimal type](https://github.com/apache/kafka/blob/9628c1278e5dcd4f6995c89809461d350d9a530a/connect/api/src/main/java/org/apache/kafka/connect/data/Decimal.java#L73) logic).

So TL;DR: it's not that we have to update that case to handle `ByteBuffer` instances (although we probably should anyways), it's that we also have to add logic that accepts `BigDecimal` instances when the schema is the logical decimal type.

I have a commit that does this exact thing; I can push it to this branch if you're interested (and feel free to revert later if you'd like).

One thing I did notice when testing this is that it doesn't appear that range properties (i.e., min/max) are supported by the generator for logical decimal types. So although we could change price to a decimal type with a scale of two, we couldn't prevent it from having negative values, for example.

Comment (ybyzek) [reply src/main/resources/productv2.avro:45]: @C0urante thank you for going down the rabbit hole.

> we couldn't prevent it from having negative values

This is a problem for the `price` field and is a deal breaker for this use case, so presumably we'll have to stick with the current implementation of this PR with price being a `long`.

IIUC the issues you highlighted could be fixed regardless.  Would it be possible for you to file an issue for tracking so we don't lose this idea?

Comment (C0urante) [reply src/main/resources/productv2.avro:45]: Sure, done (linked ticket to you offline).

Comment (C0urante) [reply src/main/resources/productv2.avro:45]: Got a PR up to add this to version 0.4.0 of the generator: https://github.com/confluentinc/avro-random-generator/pull/21

Comment (C0urante) [reply src/main/resources/productv2.avro:45]: PR's been merged, I think we can make this work with the new [0.4.0 release](https://github.com/confluentinc/avro-random-generator/tree/v0.4.0) of the generator.

File: src/main/resources/transactions.avro (hunk 4 above)
Comment (ybyzek) [reply src/main/resources/transactions.avro:26]: This is consistent with the other schemas:

```
pageviews_schema.avro:                        "regex": "User_[1-9]{0,1}"
stock_trades_schema.avro:            "regex": "User_[1-9]{0,1}"
transactions.avro:                        "regex": "User_[1-9]{0,1}"
users_array_map_schema.avro:                        "regex": "User_[1-9]{0,1}"
users_schema.avro:                        "regex": "User_[1-9]{0,1}"
```

File: src/main/resources/productv2.avro [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+                           "game",
+                           "towels"
+                         ]
+                       }
+                }},
+                {"name": "description", "type": {
+                    "type": "string",
+                    "arg.properties": {
+                      "iteration": {
+                          "start": 0
```
[the remark this answers is not in the document]
Comment (ybyzek) [reply src/main/resources/productv2.avro:41]: Yeah.  This was a straight copy & paste, but it's pretty lame.  Removing `description`

**Review by C0urante — APPROVED** [review]
LGTM

File: src/main/resources/transactions.avro (hunk 4 above)
Comment (bellemare) [reply src/main/resources/transactions.avro:26]: Yup, fine with me. I just know that this was one area that Bill and I had been talking about. Something to consider for the future, if the need arises.

File: src/main/resources/stores.avro [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        "name": "stores",
+        "type": "record",
+        "fields": [
+                {
+                  "name": "store_id",
+                  "type": "int"
+                },
+                {
+                  "name": "ein",
+                  "type": "string"
```
[the remark this answers is not in the document]
Comment (bellemare) [reply src/main/resources/stores.avro:12]: Also, what do we use ein for w.r.t. other datasets? I did a search on "ein" on this page and it seems like it only shows up for stores. What is its purpose?

**Review by bellemare — CHANGES_REQUESTED** [review]
Curious about EIN - if nothing else, I would like to see a schema "doc" tag explaining what ein is.

File: src/main/resources/stores.avro (hunk 7 above)
[the remark this answers is not in the document]
Comment (ybyzek) [reply src/main/resources/stores.avro:12]: @bellemare `EIN` is Employer Identification Number.  It's not used in other datasets, but it's teeing up use cases for "masking".

[the remark this answers is not in the document]
Comment (ybyzek) [reply src/main/resources/stores.avro:12]: @bellemare now that I think about it, the same chain would probably have the same EIN, not different ones.  I can nuke this field, WDYT?

[the remark this answers is not in the document]
Comment (ybyzek) [reply src/main/resources/stores.avro]: EIN is now removed

**Review by bellemare — APPROVED** [review]
LGTM
